src/OrchardCore/OrchardCore.Environment.Shell/ShellHost.cs

```diff
@@ -89,6 +89,38 @@ public ShellContext GetOrCreateShellContext(ShellSettings settings)
             return shell;
         }
 
+        public IServiceScope EnterServiceScope(ShellSettings settings)
+        {
+            return EnterServiceScope(settings, out var context);
+        }
+
+        public IServiceScope EnterServiceScope(ShellSettings settings, out ShellContext context)
+        {
+            context = GetOrCreateShellContext(settings);
```

Comment:
I suppose the `ShellContext` might be released immediately after this method is called (and before `TryEnterServiceScope` has a chance to be invoked).

Should `ShellHost` use good old `locks` for all its operations?

Reply:
Still working on it, just restarted but always need time to re-put all of this into my mind ;)

But the idea is to keep the fastest path in most of the cases. Then, yes the shell context might be released and `TryEnterServiceScope()` would return false (but without failing).

Then, if so we remove the shell from the concurrent dictionary and re-create it, as we now do in `GetOrCreateShellContext()` (see note 1). This is here that the atomic operation occurs in the valueFactory of the lazy value and where we create the scope.

But we will never be able to do this if the tenant is already **disabled**, the created shell would be not released but with a null service provider (see note 2), that's why we check this just before.

**Note 1**: Before, when a shell was released it was removed from the concurrent dictionary, but now a dependent shell might be released and still in the dictionary. Now in `GetOrCreateShellContext()`, we take care of this by removing and re-creating it through the value factory of the lazy value.

**Note 2**: When you get a shell settings, as long as it is not hold by an active shell context, it's state is not changed, but if it is already **disabled** `GetOrCreateShellContext()` will return a context wich is not released but with a null `ServiceProvider`.

**Note 3**: But if you already are in an active scope and the shell is **disabled** after, it is marked as released and removed from the shells dico. But you can still ends your scope with a non null provider. But on the next `GetOrCreateShellContext()` the shell will be re-created as described above.